libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthSessionRefresher.m

```diff
@@ -95,13 +96,35 @@ - (void)completeWithSuccess:(SFOAuthCredentials *)credentials {
 
 - (void)completeWithError:(NSError *)error {
     [self log:SFLogLevelError format:@"%@ Refresh failed with error: %@", NSStringFromSelector(_cmd), error];
-    if (self.errorBlock) {
+
+    if ([error.domain isEqualToString:kSFOAuthErrorDomain] && error.code == kSFOAuthErrorInvalidGrant) {
+        [self log:SFLogLevelInfo format:@"%@ Invalid grant error received, triggering logout.", NSStringFromSelector(_cmd)];
+        // make sure we call logoutUser on main thread
+        dispatch_async(dispatch_get_main_queue(), ^{
+            [self createAndStoreLogoutEvent:error];
+            [self logoutUser];
+        });
+    }
+    else if (self.errorBlock) {
         dispatch_async(dispatch_get_main_queue(), ^{
             self.errorBlock(error);
         });
     }
 }
 
+- (void)logoutUser {
+    SFUserAccountIdentity *userIdentity = [[SFUserAccountIdentity alloc] initWithUserId:self.coordinator.credentials.userId orgId:self.coordinator.credentials.organizationId];
```

Comment:
Sorry, I fell off the map yesterday before I could hit Send. 😐  This is what I meant to say!  I'll take a look at what has transpired since then.

----

> Wouldn't it be nice for the session refresher to handle the results of that

I think it could be nice for a single access point to handle that, yes.  But doing it in this manner short-circuits the error handling process of the upstream consumer, yanking the user context out from under it before it has a chance to clean up.

We used to do things more like this in iOS before, and kept running into issues where a standard error flow would dead-end in this case, leaving unfinished processes hanging around with a lack of state.

So if we can convenience-consolidate this functionality at a higher level (like `SFAuthenticationManager`, for example) and make it easier for consumers to call, I'm all for that.  But I think it's otherwise disjointed to the current design pattern to logout in the middle of an error flow, as opposed to the end of it.